src/view/renderer.js

```diff
@@ -752,6 +755,31 @@ export default class Renderer {
 
 		domSelection.collapse( anchor.parent, anchor.offset );
 		domSelection.extend( focus.parent, focus.offset );
+
+		// The following is a Firefox–specific hack (https://github.com/ckeditor/ckeditor5-engine/issues/1439).
+		// When the native DOM selection is at the end of the block and preceded by <br /> e.g.
+		//
+		//		<p>foo<br/>[]</p>
+		//
+		// which happens a lot when using the soft line break, the browser fails to (visually) move the
+		// caret to the new line. A quick fix is as simple as force–refreshing the selection with the same range.
+		if ( env.isGecko ) {
+			const parent = focus.parent;
```

Comment:
I'll move the entire body of this condition to another function because it looks bad that there's a `return` inside it. What if someone would like to add another such hack after this one? We'll need to refactor it anyway.